src/language-js/needs-parens.js

```diff
@@ -626,6 +626,14 @@ function needsParens(path, options) {
 
     case "OptionalMemberExpression":
     case "OptionalCallExpression":
+      if (
+        node.type === "OptionalMemberExpression" &&
+        parent.type === "TSNonNullExpression" &&
+        path.getParentNode(1).type === "MemberExpression"
```

Comment:
Member expressions have two child nodes, so it's not enough to check only `type` here.

Reply:
This?

```ts
(a?.b)![a?.b!]
```

Reply:
yep